File: taskManagerController/src/test/java/com/example/taskManagerController/service/TaskServiceTest.java
package com.example.taskManagerController.service;

import com.example.taskManagerController.model.Task;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.List;
import java.util.Optional;
import java.time.LocalDate;

public class TaskServiceTest {

    private TaskService taskservice;

    @BeforeEach
    void setUp() {
        taskservice = new TaskService();
    }

    // --- READ (getAllTasks) ---
    @Test
    void testGetAllTasks_returnsAllInitialTasks() {
        List<Task> allTasks = taskservice.getAllTasks();

        assertNotNull(allTasks, "List of tasks should not be null");
        assertEquals(5, allTasks.size(), "Initial task lists should be 5 tasks");
        assertEquals("Hiking", allTasks.getFirst().getTitle(), "First task title should be hiking");
    }

    // --- READ (getTaskById) ---
    @Test
    void testGetTaskById_found_returnsCorrectTask() {
        Long taskId = 2L;

        Optional<Task> result = taskservice.getTaskById(taskId);

        assertTrue(result.isPresent(), "Task with ID 2 should be found");
        assertEquals("Camping", result.get().getTitle(), "Task title should be Camping");
        assertEquals(LocalDate.of(2025, 6, 6), result.get().getDate(), "Date should be 6 June 2025");
    }

    @Test
    void testGetTaskById_notFound_returnsEmptyOptional() {
        Long nonExistentTaskId = 99L;

        Optional<Task> result = taskservice.getTaskById(nonExistentTaskId);

        assertFalse(result.isPresent(), "Task with ID 99 should not be found");
    }

    @Test
    void testGetTaskById_nullId_throwsIllegalArgumentException() {
        assertThrows(IllegalArgumentException.class, () -> taskservice.getTaskById(null),
        "Calling getTaskById with null ID should throw IllegalArgumentException");
    }

    // --- CREATE (addTask) ---
    @Test
    void testAddTask_success_increasesSizeAndIsRetrievable() {
        int initialSize = taskservice.getAllTasks().size();
        Task newTask = new Task(null, "Groceries", LocalDate.of(2025, 6, 15), "Tokyu Store", false);

        Task createdTask = taskservice.addTask(newTask);

        assertEquals(initialSize + 1, taskservice.getAllTasks().size(), "Task list size should increase by 1");
        assertEquals("Groceries", createdTask.getTitle());

        Optional<Task> foundTask = taskservice.getTaskById(createdTask.getId());
        assertTrue(foundTask.isPresent(), "Newly added task should be retrievable by its ID");
        assertEquals(createdTask.getId(), foundTask.get().getId(), "Retrieved task ID should match created task ID");
    }

    @Test
    void testAddTask_nullTask_throwsIllegalArgumentException()  {
        assertThrows(IllegalArgumentException.class, () -> taskservice.addTask(null),
        "Adding a null task should throw IllegalArgumentException");
    }

    // --- UPDATE (updateTask) ---
    @Test
    void testUpdateTask_success_updatesTaskAndReturnsOptional(){
        Long taskId = 5L;
        Task updatedTask = new Task(5L, "Cook", LocalDate.of(2025, 6, 9), "Breakfast, Lunch", false);

        Optional<Task> updatedResult = taskservice.updateTask(taskId, updatedTask);

        assertTrue(updatedResult.isPresent(), "Updated task should be present");
        assertEquals("Breakfast, Lunch", updatedResult.get().getDescription(), "Description should be updated");
    }

    void testUpdateTask_notFound_returnsEmptyOptional() {
        Long nonExistentTaskId = 99L;
        Task updatedTask = new Task(nonExistentTaskId, "Non Existent", LocalDate.of(2025, 6, 10), "Random", false);

        Optional<Task> result = taskservice.updateTask(nonExistentTaskId, updatedTask);

        assertFalse(result.isPresent(), "Updating a non-existent task should return an empty Optional");
    }

    @Test
    void testUpdateTask_nullId_throwsIllegalArgumentException() {
        Task updatedTask = new Task(null, "Cook", LocalDate.of(2025, 6, 9), "Breakfast, Lunch", false);
        assertThrows(IllegalArgumentException.class, () -> taskservice.updateTask(null, updatedTask),
        "Calling updateTask with null ID should throw IllegalArgumentException");
    }

    // --- DELETE (deleteTask) ---
    @Test
    void testDeleteTask_idFound_removesTaskAndReturnsTrue(){
        Long taskId = 4L;
        int initialSize = taskservice.getAllTasks().size();

        boolean deleted = taskservice.deleteTask(taskId);

        assertTrue(deleted, "Task with ID 4 should be successfully deleted");
        assertEquals(initialSize - 1, taskservice.getAllTasks().size(), "Task list size should decrease by 1");
    }

    @Test
    void testDeleteTask_idNotFound_returnsFalseAndKeepsSize(){
        Long taskId = 99L;
        int initialSize = taskservice.getAllTasks().size();

        boolean deleted = taskservice.deleteTask(taskId);

        assertFalse(deleted, "Deleting a non-existent task should return false");
        assertEquals(initialSize, taskservice.getAllTasks().size(), "Task list size should not change when task not found");
    }

    @Test
    void testDeleteTask_nullId_throwsIllegalArgumentException() {
        assertThrows(IllegalArgumentException.class, () -> taskservice.deleteTask(null),
        "Calling deleteTask with null ID should throw IllegalArgumentException");
    }
}

File: taskManagerController/src/test/kotlin/com/example/taskManagerController/service/TaskServiceTest.kt
package com.example.taskManagerController.service;

import com.example.taskManagerController.model.Task;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Assertions.*;
import java.time.LocalDate;
import java.util.UUID;

public class TaskServiceTest {

    private lateinit var taskService: TaskService

    private val ID_CAMPING = UUID.fromString("e72f3a8c-9d1e-4f0a-b1c2-3d4e5f6a7b8c")
    private val ID_WORK = UUID.fromString("1a2b3c4d-5e6f-7a8b-9c0d-1e2f3a4b5c6d")
    private val ID_COOK = UUID.fromString("2a3b4c5d-6e7f-8a9b-0c1d-2e3f4a5b6c7d")

    @BeforeEach
    fun setUp() {
        taskService = TaskService();
    }

    // --- READ (getAllTasks) ---
    @Test
    fun testGetAllTasks_returnsAllInitialTasks() {
        val allTasks = taskService.getAllTasks();

        assertNotNull(allTasks, "List of tasks should not be null");
        assertEquals(5, allTasks.size, "Initial task lists should be 5 tasks");
        assertEquals("Hiking", allTasks.first().title, "First task title should be hiking");
    }

    // --- READ (getTaskById) ---
    @Test
    fun testGetTaskById_found_returnsCorrectTask() {
        val result = taskService.getTaskById(ID_CAMPING);

        assertNotNull(result, "Task with ID $ID_CAMPING should be found");
        assertEquals("Camping", result?.title, "Task title should be Camping");
        assertEquals(LocalDate.of(2025, 6, 6), result?.date, "Date should be 6 June 2025")
    }

    @Test
    fun testGetTaskById_notFound_returnsEmptyOptional() {
        val nonExistentTaskId = UUID.randomUUID();
        val result = taskService.getTaskById(nonExistentTaskId);

        assertNull(result, "Task should not be found, should return null")
    }

//     --- CREATE (addTask) ---
    @Test
    fun testAddTask_success_increasesSizeAndIsRetrievable() {
        val initialSize = taskService.getAllTasks().size;
        val fakeId = UUID.randomUUID()
        val newTask = Task(fakeId,"Groceries", LocalDate.of(2025, 6, 15), "Tokyu Store", false);
        val createdTask = taskService.addTask(newTask);

        assertEquals(initialSize + 1, taskService.getAllTasks().size, "Task list size should increase by 1");
        assertEquals("Groceries", createdTask.title);

        val foundTask = taskService.getTaskById(createdTask.id);
        assertNotNull(foundTask, "Newly added task should be retrievable by its ID");
        assertEquals(createdTask.id, foundTask?.id, "Retrieved task ID should match created task ID")
    }

    // --- UPDATE (updateTask) ---
    @Test
    fun testUpdateTask_success_updatesTaskAndReturnsOptional(){
        val updatedTask = Task (ID_COOK, "Cook", LocalDate.of(2025, 6, 9), "Breakfast, Lunch", false);
        val updatedResult = taskService.updateTask(ID_COOK, updatedTask);

        assertNotNull(updatedResult, "Updated task should be present");
        assertEquals("Breakfast, Lunch", updatedResult?.description, "Description should be updated")
    }

    @Test
    fun testUpdateTask_notFound_returnsEmptyOptional() {
        val nonExistentTaskId = UUID.randomUUID();
        val updatedTask = Task(nonExistentTaskId, "Non Existent", LocalDate.of(2025, 6, 10), "Random", false);

        val result = taskService.updateTask(nonExistentTaskId, updatedTask);

        assertNull(result, "Updating a non-existent task should return an empty Optional");
    }

    // --- DELETE (deleteTask) ---
    @Test
    fun testDeleteTask_idFound_removesTaskAndReturnsTrue(){
        val initialSize = taskService.getAllTasks().size;

        val deleted = taskService.deleteTask(ID_WORK);

        assertTrue(deleted, "Task should be successfully deleted");
        assertEquals(initialSize - 1, taskService.getAllTasks().size, "Task list size should decrease by 1");
    }

    @Test
    fun testDeleteTask_idNotFound_returnsFalseAndKeepsSize(){
        val taskId = UUID.randomUUID();
        val initialSize = taskService.getAllTasks().size;

        val deleted = taskService.deleteTask(taskId);

        assertFalse(deleted, "Deleting a non-existent task should return false");
        assertEquals(initialSize, taskService.getAllTasks().size, "Task list size should not change when task not found");
    }
}
